### PR TITLE
got rid of ClusterEventReg generate ClusterEvent objects on the fly

### PR DESCRIPTION
**Pull request by minbaev**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
With this pr we don't need to lookup ClusterEvents in the ClusterEvent Regestry but rather generate these objects dynamically

#### Which issue(s) this PR fixes:
Fixes # <https://github.com/kubernetes/kubernetes/issues/99966>

#### Does this PR introduce a user-facing change?
```
NONE
```
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```
NONE
```

### CODE REVIEW

File: pkg/scheduler/eventhandlers.go [hunk 1]
```diff
@@ -119,7 +119,7 @@ func (sched *Scheduler) updateNodeInCache(oldObj, newObj interface{}) {
 	}
 
 	// Only requeue unschedulable pods if the node became more schedulable.
-	if event := nodeSchedulingPropertiesChange(newNode, oldNode); event != "" {
+	if event := nodeSchedulingPropertiesChange(newNode, oldNode); event != (framework.ClusterEvent{}) {
```
Comment (Huang-Wei) [inline pkg/scheduler/eventhandlers.go:122]: nit: better to check `if event := ...; event.Resource != ""`.

File: pkg/scheduler/framework/types.go [hunk 2]
```diff
@@ -86,6 +109,11 @@ type ClusterEvent struct {
 	ActionType ActionType
 }
 
+// String returns a string represerntation of ClusterEvent object
+func (ce ClusterEvent) String() string {
+	return fmt.Sprintf("Resource: %s, ActionType: %s", ce.Resource, ce.ActionType.String())
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/types.go:114]: As the string will be used in metrics, I'd suggest to shorten it, like:

```go
if event.Resource == "" {
    return "InvalidEvent" 
}
return fmt.Sprintf("%s:%s", ce.Resource, ce.ActionType.String())
```

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 3]
```diff
@@ -362,30 +362,43 @@ func TestPriorityQueue_Delete(t *testing.T) {
...
 		},
 		{
 			name:      "worst",
-			moveEvent: NodeAdd,
+			moveEvent: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add},
 		},
 		{
 			name: "random",
-			// leave "moveEvent" unspecified
+			// leave "golang array " unspecified
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:380]: The wording "move event" is fine.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 4]
```diff
@@ -443,7 +456,7 @@ func BenchmarkMoveAllToActiveOrBackoffQueue(b *testing.B) {
 					}
 
 					b.StartTimer()
-					if tt.moveEvent != "" {
+					if tt.moveEvent != (framework.ClusterEvent{}) {
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:459]: Ditto. `if tt.moveEvent.Resource != ""`

**Review by Huang-Wei — COMMENTED** [review]
Thanks @minbaev. Some comments.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 3 above)
Comment (minbaev) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:380]: yeah that's my typo

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 5]
```diff
@@ -1558,7 +1571,7 @@ func TestIncomingPodsMetrics(t *testing.T) {
 				moveAllToActiveOrBackoffQ,
 			},
 			want: ` scheduler_queue_incoming_pods_total{event="ScheduleAttemptFailure",queue="unschedulable"} 3
-            scheduler_queue_incoming_pods_total{event="UnschedulableTimeout",queue="backoff"} 3
+            scheduler_queue_incoming_pods_total{event="*:All",queue="backoff"} 3
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:1574]: @ahg-g This indicates there will be new event labels in metric records.

**Comment by Huang-Wei** [comment]
@minbaev could you update the release-notes section quoted? (not a plain NONE)

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 6]
```diff
@@ -389,7 +389,7 @@ func (p *PriorityQueue) flushUnschedulableQLeftover() {
 	}
 
 	if len(podsToMove) > 0 {
-		p.movePodsToActiveOrBackoffQueue(podsToMove, UnschedulableTimeout)
+		p.movePodsToActiveOrBackoffQueue(podsToMove, framework.WildCardEvent)
```
[the remark this answers is not in the document]
Comment (minbaev) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: @ahg-g I wonder how we differentiate the string representation of  `UnschedulableTimeout` and `WildCardEvent` event since the semantics is identical and String() method of `ClusterEvent` is based on this semantics

```
// String returns a string represerntation of ClusterEvent object
func (ce ClusterEvent) String() string {
	if ce.Resource == "" {
		return "InvalidEvent"
	}
	return fmt.Sprintf("%s:%s", ce.Resource, ce.ActionType.String())
}
```

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: @minbaev is right. Unless we change the stringify implementation (for now it's implemented as `Resource:ActionType`, we will still see new labels showing in the metrics.

So not quite sure it's worth making the string representation (label of metrics) backwards compatible.

[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Sorry, I meant a new resource type with the same semantics as `WildCard`. But that will be error prone.

> So not quite sure it's worth making the string representation (label of metrics) backwards compatible.

It is more about the readability of the label. All:* isn't useful. 

Here is another suggestion: 

Change `ClusterEvent` to include a `Label` field, and lets define constants for all events we are using and was listed in pkg/scheduler/internal/queue/events.go:

```
var PodAdd = ClusterEvent{Resource: Pod, ActionType: Add, Label: "PodAdd"}
...

var UnschedulableTimeout = ClusterEvent{Resource: WildCard, ActionType: All, Label: "UnschedulableTimeout"}
```

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: Adding an explicit label looks good to me. And we need to add a helper function called `IsWildcardEvent()` to compare two events on their {Resource, ActionType}, to replace current occurrences of "if event == framework.WildCardEvent":

```
pkg/scheduler/internal/queue/scheduling_queue.go
868:	if clusterEvent == framework.WildCardEvent {
879:		evtMatch := evt == framework.WildCardEvent ||
```

File: pkg/scheduler/framework/types.go [hunk 7]
```diff
@@ -98,25 +75,96 @@ const (
 	WildCard              GVK = "*"
 )
 
+// AssignedPodAdd is the event when a pod is added that causes pods with matching affinity terms
+// to be more schedulable.
+var AssignedPodAdd = ClusterEvent{Resource: Pod, ActionType: Add, Label: "AssignedPodAdd"}
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/types.go:80]: could you format the variables to be:

```go
var (
    // comment for foo
    foo = 
    // comment for bar
    bar = 
)
```

And no need to add an empty line in between.

File: pkg/scheduler/framework/types.go [hunk 8]
```diff
@@ -98,25 +75,96 @@ const (
...
+	Label      string
 }
 
-// String returns a string represerntation of ClusterEvent object
-func (ce ClusterEvent) String() string {
-	if ce.Resource == "" {
-		return "InvalidEvent"
+//IsWildCard returns true if ClusterEvent follows WildCard semantics
+func (ce ClusterEvent) IsWildCard() bool {
+	if ce.Resource == "*" && ce.ActionType == All {
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/types.go:164]:
```suggestion
	if ce.Resource == WildCard && ce.ActionType == All {
```

**Review by Huang-Wei — COMMENTED** [review]
Some nits. LGTM otherwise.

File: pkg/scheduler/framework/types.go [hunk 9]
```diff
@@ -75,80 +75,59 @@ const (
...
+	// CSINodeUpdate is the event when a CSI node is updated in the cluster.
+	CSINodeUpdate = ClusterEvent{Resource: CSINode, ActionType: Update, Label: "CSINodeUpdate"}
+	// ServiceAdd is the event when a service is added in the cluster.
+	ServiceAdd = ClusterEvent{Resource: Service, ActionType: Add, Label: "ServiceAdd"}
+	// ServiceUpdate is the event when a service is updated in the cluster.
+	ServiceUpdate = ClusterEvent{Resource: Service, ActionType: Update, Label: "ServiceUpdate"}
+	// ServiceDelete is the event when a service is deleted in the cluster.
+	ServiceDelete = ClusterEvent{Resource: Service, ActionType: Delete, Label: "ServiceDelete"}
+	// PodAdd is the event when a new pod is added to API server.
+	PodAdd = ClusterEvent{Label: "PodAdd"}
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/types.go:121]: Let's leave `PodAdd`, `ScheduleAttemptFailure` and `BackoffComplete ` as plain strings. Put them in a separate block.

File: pkg/scheduler/framework/types.go [hunk 10]
```diff
@@ -75,15 +75,75 @@ const (
 	WildCard              GVK = "*"
 )
 
-// WildCardEvent semantically matches all resources on all actions.
-var WildCardEvent = ClusterEvent{Resource: WildCard, ActionType: All}
+var (
```
Comment (ahg-g) [inline pkg/scheduler/framework/types.go:78]: I think we should put those in `pkg/scheduler/internal/queue/events.go` since they are not used by the plugins.

File: test/e2e/scheduling/events.go [hunk 11]
```diff
@@ -23,7 +23,7 @@ import (
 	"sync"
 	"time"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (ahg-g) [inline test/e2e/scheduling/events.go:26]: revert this change

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 12]
```diff
@@ -507,23 +507,23 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
...
+	p.movePodsToActiveOrBackoffQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod), framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Add})
 	p.lock.Unlock()
 }
 
 // AssignedPodUpdated is called when a bound pod is updated. Change of labels
 // may make pending pods with matching affinity terms schedulable.
 func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
 	p.lock.Lock()
-	p.movePodsToActiveOrBackoffQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod), AssignedPodUpdate)
+	p.movePodsToActiveOrBackoffQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod), framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Update})
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:518]: don't we have a constant for this event? `AssignedPodUpdate`?

Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue.go:518]: Yes, we have - `framework.AssignedPodUpdate`.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 13]
```diff
@@ -507,23 +507,23 @@ func (p *PriorityQueue) Delete(pod *v1.Pod) error {
 // may make pending pods with matching affinity terms schedulable.
 func (p *PriorityQueue) AssignedPodAdded(pod *v1.Pod) {
 	p.lock.Lock()
-	p.movePodsToActiveOrBackoffQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod), AssignedPodAdd)
+	p.movePodsToActiveOrBackoffQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod), framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Add})
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue.go:510]:
```suggestion
	p.movePodsToActiveOrBackoffQueue(p.getUnschedulablePodsWithMatchingAffinityTerm(pod), framework.AssignedPodAdded)
```

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 14]
```diff
@@ -44,6 +44,10 @@ const queueMetricMetadata = `
 		# TYPE scheduler_queue_incoming_pods_total counter
 	`
 
+var TestEvent = framework.ClusterEvent{Resource: "test"}
+var NodeAllEvent = framework.ClusterEvent{Resource: framework.Node, ActionType: framework.All}
+var EmptyEvent = framework.ClusterEvent{}
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:49]: nit: group them together:

```
var (
    TestEvent ...
    ...
)
```

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 15]
```diff
@@ -384,8 +389,22 @@ func BenchmarkMoveAllToActiveOrBackoffQueue(b *testing.B) {
...
+		framework.NodeConditionChange,
+		framework.NodeLabelChange,
+		framework.PodAdd,
+		framework.PvcAdd,
+		framework.PvcUpdate,
+		framework.PvAdd,
+		framework.PvUpdate,
+		framework.StorageClassAdd,
+		framework.CSINodeAdd,
+		framework.CSINodeUpdate}
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:406]: nit:

```suggestion
		framework.CSINodeUpdate,
	}
```

**Comment by Huang-Wei** [comment]
Thanks @minbaev !

/lgtm
/approve

/hold
in case @ahg-g wants to take another look.

File: pkg/scheduler/internal/queue/events.go [hunk 16]
```diff
@@ -20,80 +20,59 @@ import (
...
-// Events that trigger scheduler queue to change.
 const (
 	// PodAdd is the event when a new pod is added to API server.
-	PodAdd = "PodAdd"
-	// NodeAdd is the event when a new node is added to the cluster.
-	NodeAdd = "NodeAdd"
+	PodAdd string = "PodAdd"
 	// ScheduleAttemptFailure is the event when a schedule attempt fails.
-	ScheduleAttemptFailure = "ScheduleAttemptFailure"
+	ScheduleAttemptFailure string = "ScheduleAttemptFailure"
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/events.go:27]:
```suggestion
	ScheduleAttemptFailure = "ScheduleAttemptFailure"
```

**Comment by ahg-g** [comment]
one nit.

/lgtm

**Comment by ahg-g** [comment]
thanks, pls squash.

**Comment by minbaev** [comment]
@resouer can you please review?

**Comment by Huang-Wei** [comment]
> @resouer can you please review?

We don't need that. This PR is merge-ready. It will be merged once code freeze for 1.21 gets lifted.

**Comment by Huang-Wei** [comment]
> PR needs rebase.

Some other files were checked in, and caused some conflicts. Would you mind rebasing this PR?

**Comment by minbaev** [comment]
> > PR needs rebase.
> 
> Some other files were checked in, and caused some conflicts. Would you mind rebasing this PR?

Sure, will work on this

File: pkg/scheduler/scheduler_test.go [hunk 17]
```diff
@@ -719,7 +719,7 @@ func TestSchedulerFailedSchedulingReasons(t *testing.T) {
 	queuedPodStore := clientcache.NewFIFO(clientcache.MetaNamespaceKeyFunc)
 	scache := internalcache.New(10*time.Minute, stop)
 
-	// Design the baseline for the pods, and we will make nodes that don't fit it later.
+	// Design the baseline for the pods, and we will make nodes that dont fit it later.
```
Comment (ahg-g) [inline pkg/scheduler/scheduler_test.go:722]: please revert, this is not correct.

File: pkg/scheduler/eventhandlers_test.go [hunk 18]
```diff
@@ -23,7 +23,7 @@ import (
 	"time"
 
 	"github.com/google/go-cmp/cmp"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (Huang-Wei) [inline pkg/scheduler/eventhandlers_test.go:26]: Would you mind revert this change? As we're not touching this file at all.

File: pkg/scheduler/framework/types.go [hunk 19]
```diff
@@ -75,15 +75,21 @@ const (
...
 // ClusterEvent abstracts how a system resource's state gets changed.
 // Resource represents the standard API resources such as Pod, Node, etc.
 // ActionType denotes the specific change such as Add, Update or Delete.
 type ClusterEvent struct {
 	Resource   GVK
 	ActionType ActionType
+	Label      string
+}
+
+//IsWildCard returns true if ClusterEvent follows WildCard semantics
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/types.go:87]: Comment should leave a space after `//`.
